app/solvers/exact_solver.py

```diff
@@ -105,63 +59,106 @@ def duplicate(items):
-            if cities_list[krok].visited == False:
-                zysk = zysk + cities_list[krok].value
-                cities_list[krok].visited = True
-        podsumowanie[zysk] = j
+    for j in resources:
+        profit = 0
+        for stepOne in j:
+            if cities_list[stepOne].visited == False:
+                profit = profit + cities_list[stepOne].value
+                cities_list[stepOne].visited = True
+        summary[profit] = j
         g = g + 1
-        for krok in j:
-            cities_list[krok].visited = False
-        # print(i)
-        i = i + 1
-    return podsumowanie
+        for stepTwo in j:
+            cities_list[stepTwo].visited = False
+    # print(summary)
+    return summary
 
 
-def return_the_best_value(dict):
-    naj = max(dict.keys())
+# returns the best profit from dictionary which consist of paths with profits
+def return_the_best_value(dict_temp):
+    naj = max(dict_temp.keys())
     return naj
 
 
-def return_cost_of_the_best_path(dict, best):
+# return the cost on the best (from profit side) path
+def return_cost_of_the_best_path(graph, dict_temp_two, best):
     w = 1
-    koszt = 0
-    do = len(dict[best]) - 1
-    for x in dict[best][:do]:
-        koszt = koszt + dict[x][dict[best][w]]
+    cost = 0
+    do = len(dict_temp_two[best]) - 1
+    for x in dict_temp_two[best][:do]:
+        cost = cost + graph[x][dict_temp_two[best][w]]
         w = w + 1
-    return koszt
+    return cost
 
-
-def create_answer_for_path_creation(dict, best):
+# create special form of answer for path creation
+def create_answer_for_path_creation(dict_temp_three, best):
     answer_plot = []
-    do = len(dict[best]) -1
+    do = len(dict_temp_three[best]) -1
     for x in range(do):
-        answer_plot.append((dict[best][x], dict[best][x + 1]))
+        answer_plot.append((dict_temp_three[best][x], dict_temp_three[best][x + 1]))
     return answer_plot
+
+
+def solve(cities: pd.DataFrame, paths: pd.DataFrame, time: pd.DataFrame):
+    assert isinstance(cities, pd.DataFrame), 'Wrong data format!'
+    assert isinstance(paths, pd.DataFrame), 'Wrong data format!'
+    assert isinstance(time, pd.DataFrame), 'Wrong data format!'
+
+    # finds all paths from chosen start point with eye on possible time
+    def find_all_possible_paths(graph, start, time_left, path=[], price=0):
+        nonlocal all_paths
+        time_left = time_left - price
+        path = path + [start]
+        if start not in graph:
+            return []
+        for node in graph[start].keys():
+            if time_left - graph[start][node] >= 0:
+                find_all_possible_paths(graph, node, time_left, path, graph[start][node])
+            elif path not in all_paths:
+                all_paths.append(path)
+            if time_left == 0:
+                break
+        return all_paths
+
+    # general paths creating for all possible starting points
+    def create_all_possible_paths(graph, time_at_the_beggining):
```

Comment:
Analogicznie, czemu funkcja w funkcji?

Reply:
Bo jedna z nich odwołuje się do drugiej - chciałem żeby były razem. Tak sobie wymarzyłem.

Reply:
Czy to nie rozwiązuje problemu?
```python
def find_all_possible_paths(graph, start, time_left, path=[], price=0):         
    time_left = time_left - price
    path = path + [start]
    if start not in graph:
        return []

    all_paths = []
    for node in graph[start].keys():
        if time_left - graph[start][node] >= 0:
            all_paths += find_all_possible_paths(graph, node, time_left, path, graph[start][node])
        elif path not in all_paths:
            all_paths.append(path)
        if time_left == 0:
            break
    return all_paths


def create_all_possible_paths(graph, time_at_the_beggining):
    list_of_paths = []
    for x in graph.keys():
     # all_paths = []
        list_of_paths += find_all_possible_paths(graph, x, time_at_the_beggining)
    return list_of_paths
```

Reply:
Plus `time-left-price` wygląda okropnie :D

Reply:
Tomek to jest rekurencja - w tym co napisałeś - 
1. najpierw czyścisz all_paths = []
2. potem dodajesz tam np. dwie ścieżki [[A,B],[A,C]]
3. Wchodzisz do następnej iteracji i dodajesz wierzchołek D (to tylko przykład) 
4. all_paths =  [[A,B,D],[A,C,D]]
5. a potem czyścisz all_paths=[] bo wszedłeś w następną rekurencję 
6. kończą Ci się wierzchołki i zwracasz pustą listę.

Reply:
> najpierw czyścisz all_paths = []
> potem dodajesz tam np. dwie ścieżki [[A,B],[A,C]]
> Wchodzisz do następnej iteracji i dodajesz wierzchołek D (to tylko przykład)

NIE

> all_paths = [[A,B,D],[A,C,D]] 

TYLKO
all_paths [[A,B],[A,C],[A,B,D],[A,C,D]]

> a potem czyścisz all_paths=[] bo wszedłeś w następną rekurencję
tak, ale pierwsze all_paths czeka nie zostało wyczyszczone, ono czeka az rekurencja się skończy i wtedy będzie lista wszystkich ścieżek.

> kończą Ci się wierzchołki i zwracasz pustą listę.
Sprawdź ;)

Reply:
Ogólnie wydaje mi się że Tomek ma rację jeśli chodzi o rekurencję.
Adelef, masz output, który pokazywałby, że to działa tak jak mówisz?

W funkcji:
```
def find_all_possible_paths(graph, start, time_left, path=[], price=0):
```
nazwa `price` faktycznie jest myląca - lepiej zamienić na `travel_time` czy coś w ten deseń.

Reply:
Proszę oto wynik przy starej wersji: 
" 471
Wykorzystany czas - 15
[('city_129', 'city_127'), ('city_127', 'city_129'), ('city_129', 'city_127'), ('city_127', 'city_126'), ('city_126', 'city_128'), ('city_128', 'city_360'), ('city_360', 'city_475'), ('city_475', 'city_482')]
[0, 471, ['city_129', 'city_127', 'city_129', 'city_127', 'city_126', 'city_128', 'city_360', 'city_475', 'city_482']]

Process finished with exit code 0
"

Oto wynik po zaproponowanej poprawce (wyrzucone nonlocal i dorzucone all_paths=[] w zaproponowanej wersji): 
" 100
Wykorzystany czas - 0
[]
[15, 100, ['city_466']]

Process finished with exit code 0
" 
FAKTYCZNIE DZIAŁA SZYBCIEJ. 
Rozumiem, że mam tak zostawić?

Reply:
Jeżeli chodzi o resztę - ja szanuje wszystkie ważne komentarze i poprawki, ale kwestia tego, że coś sie komuś bardziej podoba lub mniej to raczej subiektywna opinia. Wprowadzę zmiany jak najbardziej, chętnie, ale w momencie kiedy ma to jakiś związek z działaniem kodu. Z pewnością docstringi są super extra, szanuje. Ale to nie są zajęcia z programowania w Pythonie. Może trochę nie przesadzajmy - to ma działać i tyle.

Reply:
A z tą rekurencją to nie wiem - wrzuciłem zmianę i wkleiłem Wam wynik - może coś źle zmieniłem, ale tak jak ja rozumiem, jak idzie ta pętla i rekurencja to tak powinno zostać. Mogę się spotkać i na ten temat podyskutować i powiedziec dlaczego tak napisałem.